File: Interview Preparation Kit/Recursion and Backtracking/RecursiveDigitSum.js
// https://www.hackerrank.com/challenges/recursive-digit-sum/problem
'use strict';

const fs = require('fs');

process.stdin.resume();
process.stdin.setEncoding('utf-8');

let inputString = '';
let currentLine = 0;

process.stdin.on('data', inputStdin => {
    inputString += inputStdin;
});

process.stdin.on('end', function() {
    inputString = inputString.replace(/\s*$/, '')
        .split('\n')
        .map(str => str.replace(/\s*$/, ''));

    main();
});

function readLine() {
    return inputString[currentLine++];
}

// Complete the digitSum function below.
function superDigit(n, k) {
    if (n.length == 1 && k == 1) {
        return parseInt(n);
    }
    
    let sum = 0;
    for (let i = 0; i < n.length; i++) {
        sum += parseInt(n[i]) * k;
    }
    
    return superDigit(sum.toString(), 1);
}

function main() {
    const ws = fs.createWriteStream(process.env.OUTPUT_PATH);

    const nk = readLine().split(' ');

    const n = nk[0];

    const k = parseInt(nk[1], 10);

    const result = superDigit(n, k);

    ws.write(result + '\n');

    ws.end();
}

File: Interview Preparation Kit/Recursion and Backtracking/RecursiveDigitSum.go
// https://www.hackerrank.com/challenges/recursive-digit-sum/problem
package main

import (
	"bufio"
	"fmt"
	"io"
	"os"
	"strconv"
	"strings"
)

/*
 * Complete the 'superDigit' function below.
 *
 * The function is expected to return an INTEGER.
 * The function accepts following parameters:
 *  1. STRING n
 *  2. INTEGER k
 */
func superDigit(n string, k int32) int32 {
	n_len := len(n)
	if n_len == 1 && k == 1 {
		val, _ := strconv.Atoi(n)
		return int32(val)
	}

	sum := int64(0)
	for _, digit := range n {
		sum += int64(digit-'0') * int64(k)
	}

	return superDigit(strconv.FormatInt(sum, 10), 1)
}

func main() {
	reader := bufio.NewReaderSize(os.Stdin, 16*1024*1024)

	stdout, err := os.Create(os.Getenv("OUTPUT_PATH"))
	checkError(err)

	defer stdout.Close()

	writer := bufio.NewWriterSize(stdout, 16*1024*1024)

	firstMultipleInput := strings.Split(strings.TrimSpace(readLine(reader)), " ")

	n := firstMultipleInput[0]

	kTemp, err := strconv.ParseInt(firstMultipleInput[1], 10, 64)
	checkError(err)
	k := int32(kTemp)

	result := superDigit(n, k)

	fmt.Fprintf(writer, "%d\n", result)

	writer.Flush()
}

func readLine(reader *bufio.Reader) string {
	str, _, err := reader.ReadLine()
	if err == io.EOF {
		return ""
	}

	return strings.TrimRight(string(str), "\r\n")
}

func checkError(err error) {
	if err != nil {
		panic(err)
	}
}
